### PR TITLE
Fix some issues found by Coverity MISRA

### PR DESCRIPTION
**Pull request by sarenameas**
This PR fixes some outstanding issues found by Coverity MISRA. After this PR there will be only 45 rule violations in the core_http_client.c source, all of which are related to using non-compliant http-parser.

This includes:
- Making explicit casts in comparisons for size_t against a narrower type.
- Delete unnecessary comparison that was just a work around for cbmc's transport interface send stub.
- Change all memcpy to strncpy for MISRA rule 7.4.

[cov-out.zip](https://github.com/FreeRTOS/coreHTTP/files/5436854/cov-out.zip)

### CODE REVIEW

File: test/cbmc/proofs/HTTPClient_AddHeader/Makefile [hunk 1]
```diff
@@ -23,7 +23,17 @@ HARNESS_ENTRY=HTTPClient_AddHeader_harness
...
+# bounded to reduce the proof run time. This length is a bound on the strncpys
+# used to write the header field and values to the request header buffer. Memory
+# safety on the request header buffer can be proven in a reasonable bound. It
+# adds no value to the proof to write a header field and a header value of
+# SIZE_MAX length.
+HEADER_FIELD_MAX_LENGTH=16
+HEADER_VALUE_MAX_LENGTH=16
+
+DEFINES += -DHEADER_FIELD_MAX_LENGTH=$(HEADER_FIELD_MAX_LENGTH)
+DEFINES += -DHEADER_VALUE_MAX_LENGTH=$(HEADER_VALUE_MAX_LENGTH)
```
Comment (sarenameas) [inline test/cbmc/proofs/HTTPClient_AddHeader/Makefile:36]: This functions runs quickly without stubbing out strncpy.

File: test/cbmc/stubs/transport_interface_stubs.c [hunk 2]
```diff
@@ -48,14 +48,21 @@ int32_t TransportInterfaceSendStub( NetworkContext_t * pNetworkContext,
...
 
     if( tries >= MAX_TRIES )
     {
         tries = 0;
-        ret = bytesToSend;
+
+        /* In order to stop the looping on send we must return an error or
+         * bytesToSend. We return an error instead of bytesToSend because
+         * bytesToSend may be a value larger than INT32_MAX. */
+        ret = -1;
```
Comment (yourslab) [inline test/cbmc/stubs/transport_interface_stubs.c:65]: Do we always want to return an error in the last attempt though even though it is possible to send remaining bytes? Would a compromise between the two approaches work?
```suggestion
        if( bytesToSend >= INT32_MAX )
        {
            ret = -1;
        }
        else
        {
            ret = bytesToSend;
        }
```

File: test/cbmc/proofs/HTTPClient_InitializeRequestHeaders/Makefile [hunk 3]
```diff
@@ -38,13 +38,13 @@ REMOVE_FUNCTION_BODY += __CPROVER_file_local_core_http_client_c_httpParserOnHead
...
 
-REMOVE_FUNCTION_BODY += memcpy
+REMOVE_FUNCTION_BODY += strncpy
 
 UNWINDSET += strncmp.0:5
 
 PROOF_SOURCES += $(PROOFDIR)/$(HARNESS_FILE).c
 PROOF_SOURCES += $(SRCDIR)/test/cbmc/sources/http_cbmc_state.c
-PROOF_SOURCES += $(SRCDIR)/test/cbmc/stubs/memcpy.c
+PROOF_SOURCES += $(SRCDIR)/test/cbmc/stubs/strncpy.c
```
Comment (yourslab) [inline test/cbmc/proofs/HTTPClient_InitializeRequestHeaders/Makefile:47]: Is the stub only used here because the proof for this function hangs or runs too long?

File: test/cbmc/proofs/HTTPClient_AddHeader/Makefile (hunk 1 above)
Comment (yourslab) [reply test/cbmc/proofs/HTTPClient_AddHeader/Makefile:36]: Do we have a rule of thumb for when we are allowed to bound the size of buffers? The user is always allowed to pass a header field or value that is larger than that new bound. It may be trivial to prove now, but a code change can easily invalidate these proofs.

File: test/cbmc/stubs/transport_interface_stubs.c (hunk 2 above)
Comment (sarenameas) [reply test/cbmc/stubs/transport_interface_stubs.c:65]: This is not necessary because CBMC seems to have been smart enough to return the expected amount of bytes to attain coverage of that case.

File: test/cbmc/proofs/HTTPClient_AddHeader/Makefile (hunk 1 above)
Comment (sarenameas) [reply test/cbmc/proofs/HTTPClient_AddHeader/Makefile:36]: We bound everywhere that using SIZE_MAX would loop for SIZE_MAX increasing the run time of the proof. This is done in JSON, Shadow, Defender, and MQTT. Every library except this one.

File: test/cbmc/proofs/HTTPClient_InitializeRequestHeaders/Makefile (hunk 3 above)
Comment (sarenameas) [reply test/cbmc/proofs/HTTPClient_InitializeRequestHeaders/Makefile:47]: Yes. As is was in the previous iteration with memcpy.

**Review by yourslab — APPROVED** [review]

File: source/core_http_client.c [hunk 4]
```diff
@@ -1218,24 +1218,27 @@ static HTTPStatus_t addHeader( HTTPRequestHeaders_t * pRequestHeaders,
     {
         /* Write "<Field>: <Value> \r\n" to the headers buffer. */
 
-        /* Copy the header name into the buffer. */
-        ( void ) memcpy( pBufferCur, pField, fieldLen );
+        /* Copy the header name into the buffer. memcpy can be used here, but
```
Comment (gkwicker) [inline source/core_http_client.c:1221]: I would undo the change to the comment here as well as at lines 1226-1228.  You can use `strncpy()` without any additional explanation.

File: source/core_http_client.c [hunk 5]
```diff
@@ -1723,19 +1732,9 @@ static HTTPStatus_t sendHttpHeaders( const TransportInterface_t * pTransport,
...
-                        "2147483647 (INT32_MAX)." ) );
-            returnStatus = HTTPInvalidParameter;
-        }
-        else
-        {
-            LogDebug( ( "Sending HTTP request headers: HeaderBytes=%lu",
-                        ( unsigned long ) ( pRequestHeaders->headersLen ) ) );
-            returnStatus = sendHttpData( pTransport, pRequestHeaders->pBuffer, pRequestHeaders->headersLen );
-        }
+        LogDebug( ( "Sending HTTP request headers: HeaderBytes=%lu",
```
Comment (gkwicker) [inline source/core_http_client.c:1735]: Why was the check on the previous line 1727 removed?  Is that an impossible condition here?

File: source/core_http_client.c [hunk 6]
```diff
@@ -1999,7 +1998,7 @@ HTTPStatus_t HTTPClient_Send( const TransportInterface_t * pTransport,
                     "reqBodyBufLen is greater than zero." ) );
         returnStatus = HTTPInvalidParameter;
     }
-    else if( reqBodyBufLen > INT32_MAX )
+    else if( reqBodyBufLen > ( size_t ) ( INT32_MAX ) )
```
Comment (gkwicker) [inline source/core_http_client.c:2001]: Why are you including the decimal value of `INT32_MAX` in the log message at line 2005?  This value is well-defined so adding the literal value is not helpful and consumes storage space.

File: source/core_http_client.c [hunk 7]
```diff
@@ -2254,6 +2250,9 @@ static HTTPStatus_t findHeaderInResponse( const uint8_t * pBuffer,
                     pField,
                     ( int ) ( *pValueLen ),
                     *pValueLoc ) );
+
```
Comment (gkwicker) [inline source/core_http_client.c:2253]: The comment on line 2254 doesn't seem helpful, I would delete it.  This `else` block is doing something important, which is to set `returnStatus` to indicate that the requested header was found in the response.

**Review by gkwicker — CHANGES_REQUESTED** [review]
See comments

File: source/core_http_client.c (hunk 5 above)
Comment (sarenameas) [reply source/core_http_client.c:1735]: This check was removed because the condition was there to satisfy some incorrect logic in the CBMC stub. 
We should need to limit the headers we are sending to less than size_t for the network interface send function.

File: source/core_http_client.c (hunk 6 above)
Comment (sarenameas) [reply source/core_http_client.c:2001]: Sure thanks

File: source/core_http_client.c (hunk 7 above)
Comment (sarenameas) [reply source/core_http_client.c:2253]: you are right thanks

File: source/core_http_client.c (hunk 4 above)
Comment (sarenameas) [reply source/core_http_client.c:1221]: Sure thanks

**Review by gkwicker — APPROVED** [review]

**Review by yourslab — APPROVED** [review]
